fix(Gmail Node): Set References and In-Reply-To only when user provides threadId

dana-gill opened the pull request:
## Summary

Set References and In-Reply-To only when user provides threadId. Gets the latest message from the thread to do this.

We also need to adapt the `nock` for creating a draft in tests since the encoding for the email gets dynamically generated we can no longer rely on the raw message being the same on each run

Workflow for testing:
```
{
  "nodes": [
    {
      "parameters": {
        "pollTimes": {
          "item": [
            {
              "mode": "everyMinute"
            }
          ]
        },
        "filters": {}
      },
      "type": "n8n-nodes-base.gmailTrigger",
      "typeVersion": 1.2,
      "position": [
        -64,
        560
      ],
      "id": "8346fdd7-9e69-4f78-b9f2-349678d27a8e",
      "name": "Gmail Trigger1",
      "credentials": {
        "gmailOAuth2": {
          "id": "SLV3D68Mpqomofzr",
          "name": "Gmail account"
        }
      }
    },
    {
      "parameters": {
        "resource": "draft",
        "subject": "={{ $('Gmail Trigger1').item.json.Subject }}",
        "message": "=hello from n8n ",
        "options": {
          "threadId": "={{ $('Gmail Trigger1').item.json.threadId }}"
        }
      },
      "type": "n8n-nodes-base.gmail",
      "typeVersion": 2.1,
      "position": [
        416,
        688
      ],
      "id": "919d9fcd-c3fd-4f08-ad38-0afe0a29684f",
      "name": "no set, no reply to, draft",
      "webhookId": "2ae44867-684e-49d1-b1a5-c85794b129d4",
      "credentials": {
        "gmailOAuth2": {
          "id": "SLV3D68Mpqomofzr",
          "name": "Gmail account"
        }
      }
    }
  ],
  "connections": {
    "Gmail Trigger1": {
      "main": [
        [
          {
            "node": "no set, no reply to, draft",
            "type": "main",
            "index": 0
          }
        ]
      ]
    }
  },
  "pinData": {},
  "meta": {
    "templateCredsSetupCompleted": true,
    "instanceId": "ee90fdf8d57662f949e6c691dc07fa0fd2f66e1eee28ed82ef06658223e67255"
  }
}
```

## Related Linear tickets, Github issues, and Community forum posts

https://github.com/n8n-io/n8n/issues/15775
https://linear.app/n8n/issue/NODE-3052/community-issue-gmail-create-draft-to-reply-breaks-the-thread

## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

Commit:
Allow users to input References and In-Reply-To

Commit:
Adapt tests

Commit:
make typecheck pass

Commit:
Set headers automatically if threadId is set

Commit:
Fix tests

Commit:
make typecheck pass

Commit:
remove references and inreplyto

was asked to do this following code review

ShireenMissi commented on packages/nodes-base/nodes/Google/Gmail/v2/GmailV2.node.ts:
```diff
@@ -511,6 +511,9 @@ export class GmailV2 implements INodeType {
 						let fromAlias = '';
 						let threadId = null;
 
+						const references = options.references as string | undefined;
+						const inReplyTo = options.inReplyTo as string | undefined;
+
```
same as above please remove the manual setting of these fields
```suggestion
```

ShireenMissi commented on packages/nodes-base/nodes/Google/Gmail/v2/GmailV2.node.ts:
```diff
@@ -561,10 +564,21 @@ export class GmailV2 implements INodeType {
+							);
+						}
+
+						if (references && inReplyTo) {
+							email.inReplyTo = inReplyTo;
+							email.reference = references;
```
same here
```suggestion
```

ShireenMissi commented on packages/nodes-base/nodes/Google/Gmail/v2/utils/draft.ts:
```diff
@@ -4,27 +4,43 @@ import type { IEmail } from '@utils/sendAndWait/interfaces';
-		{},
-		{ format: 'metadata', metadataHeaders: ['Message-ID'] },
-	);
+	if (references && inReplyTo) {
```
same here

ShireenMissi commented on packages/nodes-base/nodes/Google/Gmail/v2/utils/draft.ts:
```diff
@@ -4,27 +4,43 @@ import type { IEmail } from '@utils/sendAndWait/interfaces';
+		)?.value;
+
+		email.inReplyTo = messageId;
+		email.reference = messageId;
```
this should be references
```suggestion
		email.references = messageId;
```

ShireenMissi commented on packages/nodes-base/nodes/Google/Gmail/v2/utils/draft.ts:
```diff
@@ -4,27 +4,43 @@ import type { IEmail } from '@utils/sendAndWait/interfaces';
+			(header: { name: string; value: string }) => header.name === 'Message-ID',
+		)?.value;
+
+		email.inReplyTo = messageId;
```
Add validation if messageId in undefined you probably shouldn't set the headers

ShireenMissi replied:
The References and In-Reply-To headers must be set in compliance with the [RFC 2822](https://tools.ietf.org/html/rfc2822#appendix-A.2) standard.
```suggestion
	email.inReplyTo = `<${messageId}>`;
```

ShireenMissi reviewed (dismissed):
Nice work on the auto-population of headers logic! I think we can simplify this by removing the manual References and In-Reply-To fields since the auto-population handles this automatically. This would reduce complexity for users. 
Please also check my inline comments for a few technical suggestions.

dana-gill commented:
@ShireenMissi sorry, i realized that i forgot to push my changes which removes references and inReplyTo from node description

Commit:
setEmailReplyHeaders

Commit:
use reference => references

Commit:
make message id more general

Commit:
fix tests

Commit:
Adapt tests

The encoding for the email gets dynamically generated we can no longer
rely on the raw message being the same

janober commented:
Got released with [`n8n@1.103.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.103.0)
